client/service/src/lib.rs

```diff
@@ -116,21 +131,101 @@ pub struct Service<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> {
 
 impl<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> Unpin for Service<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> {}
 
+/// Client super trait, use this instead of the concrete Client type.
+pub trait ClientProvider<
```

Comment:
The thought behind this and the related changes was that we'd not have a giant objects that everyone relies upon anymore but specific traits that are to be implemented for the various features. Ergo the default bound of `Service::Client` should also limited to those traits absolutely necessary for the trait (builder), not all traits that some features might require. Even if `client` does implement them all at the moment, having them bound here also means we force that architecture. We should split this up into two traits for now: the minimum and the max that implementations can  just force for the time being – but document at this is a helper and specific trait bounds should be used instead.

The patch that worked for me is something like:
```diff
index f379790a9..d92eae294 100644
--- a/client/service/src/lib.rs
+++ b/client/service/src/lib.rs
@@ -132,81 +132,94 @@ pub struct Service<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> {
 impl<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> Unpin for Service<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> {}
 
 /// Client super trait, use this instead of the concrete Client type.
-pub trait ClientProvider<
+pub trait MinimalClient<
+       Block: BlockT,
+>:
+       HeaderMetadata<Block>
+       + BlockchainEvents<Block>
+       + UsageProvider<Block>
+       + CallApiAt<Block>
+{}
+
+pub trait AllFeaturesClient<
        Block: BlockT,
        Backend: BackendT<Block>,
-       Executor: CallExecutor<Block>,
-       Runtime: ConstructRuntimeApi<Block, Self>,
+       Executor: 'static + CallExecutor<Block> + Send + Sync + Clone,
+       Runtime: ConstructRuntimeApi<Block, Self> + Send + Sync,
 >:
-       HeaderBackend<Block>
+       MinimalClient<Block>
        + ProvideRuntimeApi<
                Block,
                Api = <Runtime as ConstructRuntimeApi<Block, Self>>::RuntimeApi
        >
+       + BlockBuilderProvider<Backend, Block, Self>
+       + ExecutorProvider<Block, Executor = Executor>
+       + HeaderBackend<Block>
        + LockImportRun<Block, Backend>
        + ProofProvider<Block>
-       + BlockBuilderProvider<Backend, Block, Self>
        + ProvideUncles<Block>
        + StorageProvider<Block, Backend>
        + Chain<Block>
-       + HeaderMetadata<Block, Error = sp_blockchain::Error>
-       + ExecutorProvider<Block, Executor = Executor>
        + ProvideCache<Block>
        + BlockIdTo<Block, Error = sp_blockchain::Error>
-       + CallApiAt<
-               Block,
-               Error = sp_blockchain::Error,
-               StateBackend = <Backend as BackendT<Block>>::State
-       >
        + BlockImport<
                Block,
                Error = sp_consensus::Error,
                Transaction = TransactionFor<Backend, Block>
        >
        + Finalizer<Block, Backend>
-       + BlockchainEvents<Block>
        + BlockBackend<Block>
-       + UsageProvider<Block>
        + AuxStore
 {}
 
-impl<Block, Backend, Executor, Runtime> ClientProvider<Block, Backend, Executor, Runtime>
+impl<Block, Backend, Executor, Runtime> MinimalClient<Block>
        for
                Client<Backend, Executor, Block, Runtime>
        where
                Block: BlockT,
-               Backend: BackendT<Block>,
-               Executor: CallExecutor<Block>,
-               Runtime: ConstructRuntimeApi<Block, Self>,
-               Self: HeaderBackend<Block>
+               Backend: 'static + BackendT<Block>,
+               Self:
+                       CallApiAt<
+                               Block,
+                               Error = sp_blockchain::Error,
+                               StateBackend = <Backend as BackendT<Block>>::State
+                       >
+                       + HeaderMetadata<Block, Error = sp_blockchain::Error>
+                       + BlockchainEvents<Block>
+                       + UsageProvider<Block>
+{}
+
+impl<Block, Backend, Executor, Runtime> AllFeaturesClient<Block, Backend, Executor, Runtime>
+       for
+               Client<Backend, Executor, Block, Runtime>
+       where
+               Block: BlockT,
+               Backend: 'static + BackendT<Block>,
+               Executor: 'static + CallExecutor<Block> + Send + Sync + Clone,
+               Runtime: ConstructRuntimeApi<Block, Self> + Send + Sync,
+               Self:
+                       MinimalClient<Block>
                        + ProvideRuntimeApi<
                                Block,
                                Api = <Runtime as ConstructRuntimeApi<Block, Self>>::RuntimeApi
-                       >
+                               >
+                       + BlockBuilderProvider<Backend, Block, Self>
+                       + ExecutorProvider<Block, Executor = Executor>
+                       + HeaderBackend<Block>
                        + LockImportRun<Block, Backend>
                        + ProofProvider<Block>
-                       + BlockBuilderProvider<Backend, Block, Self>
                        + ProvideUncles<Block>
                        + StorageProvider<Block, Backend>
                        + Chain<Block>
-                       + HeaderMetadata<Block, Error = sp_blockchain::Error>
-                       + ExecutorProvider<Block, Executor = Executor>
                        + ProvideCache<Block>
                        + BlockIdTo<Block, Error = sp_blockchain::Error>
-                       + CallApiAt<
-                               Block,
-                               Error = sp_blockchain::Error,
-                               StateBackend = <Backend as BackendT<Block>>::State
-                       >
                        + BlockImport<
                                Block,
                                Error = sp_consensus::Error,
                                Transaction = TransactionFor<Backend, Block>
                        >
                        + Finalizer<Block, Backend>
-                       + BlockchainEvents<Block>
                        + BlockBackend<Block>
-                       + UsageProvider<Block>
                        + AuxStore
 {}
 
@@ -225,7 +238,7 @@ pub trait AbstractService: Future<Output = Result<(), Error>> + Send + Unpin + S
        /// Transaction pool.
        type TransactionPool: TransactionPool<Block = Self::Block> + MallocSizeOfWasm;
        /// The generic Client type
-       type Client: ClientProvider<Self::Block, Self::Backend, Self::CallExecutor, Self::RuntimeApi>;
+       type Client: MinimalClient<Self::Block> + Send + Sync;
 
        /// Get event stream for telemetry connection established events.
        fn telemetry_on_connect_stream(&self) -> TracingUnboundedReceiver<()>;

```